### PR TITLE
Update IntervalShardingAlgorithm.java to fix an exception caused by polymorphism

### PR DESCRIPTION
**Pull request by cowboysync**
the method private String getDateTimeText(final Comparable<?> endpoint) will throws an java.lang.UnsupportedOperationException if the parameter endpoint's type is java.sql.Date, because of java.sql.Date values do not have a time component, so it's mehtod public Instant toInstant() always throws an java.lang.UnsupportedOperationException.

Reproduction example: examples/shardingsphere-jdbc-example/single-feature-example/sharding-example/sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/sharding/raw/jdbc/ShardingRawYamlIntervalConfigurationExample.java

image

Changes proposed in this pull request:
- fix an exception caused by polymorphism

### CODE REVIEW

**Review by lwclover — APPROVED** [review]

**Review by linghengqian — COMMENTED** [review]
- The CI bug was fixed by https://github.com/apache/shardingsphere/pull/20274, this PR should merge the master branch.

- It is better to add a test case of `java.sql.Date` to the corresponding unit test class of this class. Previous tests missed this class.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/IntervalShardingAlgorithmTest.java [hunk 1]
```diff
@@ -448,4 +448,12 @@ public void assertIntegerInJDBCType() {
                 new RangeShardingValue<>("t_order", "create_time", DATA_NODE_INFO, Range.closed("04", "10")));
         assertThat(actualAsMonthString.size(), is(4));
     }
+    
+    @Test
+    public void assertDateInSqlDate() {
+        Collection<String> actualAsLocalDate = shardingAlgorithmByJDBCDate.doSharding(availableTablesForJDBCDateDataSources,
+                new RangeShardingValue<>("t_order", "create_time", DATA_NODE_INFO,
+                        Range.closed(new Date(2021 - 1900, 6 - 1, 15), new Date(2021 - 1900, 7 - 1, 31))));
```
Comment (linghengqian) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/IntervalShardingAlgorithmTest.java:456]: Why use a constructor that was deprecated in JDK8?  This breaks CI for JDK18.

Comment (cowboysync) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/algorithm/sharding/datetime/IntervalShardingAlgorithmTest.java]: change to use the Java 8 Date Time API to construct java.sql.Date instance

**Review by linghengqian — APPROVED** [review]

**Comment by linghengqian** [comment]
- There are too many commits for this PR, you might open a new issue and submit a new PR?

**Pull request closed without merging**